File: Matrix/Python/Spiral Matrix III.py
# Problem link: https://leetcode.com/problems/spiral-matrix-iii/description
# Time: O(m * n)
# Space: O(1)
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        result, index, steps = [], 0, 1
        direction = [[0, 1], [1, 0], [0, -1], [-1, 0]]

        while len(result) < rows * cols:
            for x in range(2):
                for y in range(steps):
                    if 0 <= rStart < rows and 0 <= cStart < cols:
                        result.append([rStart, cStart])
                    rStart, cStart = rStart + direction[index][0], cStart + direction[index][1]
                index = (index + 1) % 4
            steps += 1
        
        return result

File: Matrix/Java/Spiral Matrix III.java
// Problem link: https://leetcode.com/problems/spiral-matrix-iii/description
// Time: O(m * n)
// Space: O(1)
class Solution {
    public int[][] spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        int[][] result = new int[rows * cols][2];
        int[][] direction = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int steps = 1, index = 0, dir = 0;
        

        while(index < rows * cols) {
            for(int idx = 0; idx < 2; idx++) {
                for(int step = 0; step < steps; step++) {
                    if(rStart >= 0 && rStart < rows && cStart >= 0 && cStart < cols) {
                        result[index][0] = rStart;
                        result[index++][1] = cStart;
                    }
                    rStart += direction[dir][0];
                    cStart += direction[dir][1];
                }
                dir = (dir + 1) % 4;
            }
            steps++;
        }

        return result;
    }
}
